api/internal/git/repospec.go

```diff
@@ -227,56 +227,111 @@ func parsePath(n string) string {
 	return parsed.Path
 }
 
-func parseHostSpec(n string) (string, string) {
-	var host string
-	// Start accumulating the host part.
-	for _, p := range []string{
-		// Order matters here.
-		"git::", "gh:", "ssh://", "https://", "http://", "file://",
-		"git@", "github.com:", "github.com/"} {
-		if len(p) < len(n) && strings.ToLower(n[:len(p)]) == p {
-			n = n[len(p):]
-			host += p
+func extractHost(n string) (string, string) {
+	// We used to use go-getter to handle our urls: https://github.com/hashicorp/go-getter.
+	// This prefix signaled go-getter to use the git protocol to fetch the url's contents.
+	// We still accept this prefix.
+	n, _ = trimPrefixIgnoreCase(n, "git::")
+
+	scheme, n := extractScheme(n)
+	if scheme == "file://" || scheme == "gh:" {
+		// We support gh: assuming authors use it as a github shorthand, specified in .gitconfig.
+		// The file protocol specifies an absolute path to a local git repo. There is no host.
+		// In both cases, we return the scheme and the rest of the url.
+		return scheme, n
+	}
+
+	username, n := extractUsername(n)
+
+	if rest, isGithubURL := trimGithubHost(n); isGithubURL {
+		// we strictly normalize Github URLs; this may discard scheme and username components in some cases.
+		return normalizedGithubHost(scheme, username), rest
+	}
+
+	sepIndex := findPathSeparator(n, username == gitUsername)
+	host, rest := n[:sepIndex+1], n[sepIndex+1:]
+
+	if !validHostSpecParsed(scheme, username, host) {
+		return "", n
+	}
+	return scheme + username + host, rest
+}
+
+func validHostSpecParsed(scheme, username, host string) bool {
+	isSCP := username == gitUsername
+	return len(host) > 0 &&
+		(scheme != "" || isSCP) // scheme may be blank for URLs like git@github.com:org/repo
+}
+
+func extractScheme(s string) (string, string) {
+	for _, prefix := range []string{"gh:", "ssh://", "https://", "http://", "file://"} {
+		if rest, found := trimPrefixIgnoreCase(s, prefix); found {
+			return prefix, rest
 		}
 	}
-	if host == "git@" {
-		i := strings.Index(n, "/")
-		if i > -1 {
-			host += n[:i+1]
-			n = n[i+1:]
-		} else {
-			i = strings.Index(n, ":")
-			if i > -1 {
-				host += n[:i+1]
-				n = n[i+1:]
-			}
+	return "", s
+}
+
+func extractUsername(s string) (string, string) {
```

Comment:
Well spotted! As it happens, the code I have pending for the PR to support custom usernames with SSH on Github has this code, which I'm able to do because the intermediate PR introduces support for returning errors:

```diff
+func normalizeGithubHost(scheme, username string) (string, error) {
+       if strings.HasPrefix(scheme, "http") && username != "" {
+               return "", InvalidGitURLError{URL: scheme + username, Reason: "usernames are not supported for http(s) Github urls"}
+       }
+       if strings.HasPrefix(scheme, "ssh://") || username != "" {
+               return normalizedSCPGithub(username), nil
        }
-       return normalizedHttpGithub
+       return normalizedHTTPGithub, nil
 }
 ```

I think we can be very confident in doing that for Github, but it sounds like we should consider this validation for all http(s) targets. I'll add a TODO in my other branch.